### ISSUE TITLE
Problems with zarr 2.11.0

### ISSUE THREAD

**Issue by frezaei98** [tskit-dev/tsinfer#628]
From Tuesday,  this code "inferred_ts = tsinfer.infer(sample_data)" has not run suddenly.  I am facing this error " TypeError: 'int' object is not a mapping". What can I do to fix this error?

**Comment by jeromekelleher** [comment]
Hi @frezaei98 :wave:, thanks for the bug report.

Can you tell us what version you're using (is it git or from conda or pip)? A minimal working example would be ideal, with a small sample data file.

**Comment by hyanwong** [comment]
That is strange. Nothing has changed in the code. Do you know what line number is giving the error. Could you post a simple test case (with data) that reproduces the error for you.

(snap, Jerome, sorry!)

**Comment by frezaei98** [comment]
> Can you tell us what version you're using (is it git or from conda or pip)? A minimal working example would be ideal, with a small sample data file.

Hi, 
Thank you very much for your respond.
I am using the conda. But I also checked it in Google colab with pip .
I tried to run the toy example in tsinfer.readthedocs.io, but it has same error.

**Comment by frezaei98** [comment]
> I tried to run the toy example in tsinfer.readthedocs.io, but it has same error.

Hi,
I tried to run the toy example in tsinfer.readthedocs.io, but it has same error.

**Comment by jeromekelleher** [comment]
I can confirm this happens in a fresh venv... investigating.

**Comment by jeromekelleher** [comment]
Very strange:
```
(venv) jk@tempest$ python3 tsinfer-example.py 
Traceback (most recent call last):
  File "tsinfer-example.py", line 12, in <module>
    inferred_ts = tsinfer.infer(sample_data)
  File "/home/jk/tmp/venv/lib/python3.8/site-packages/tsinfer/inference.py", line 291, in infer
    ancestors_ts = match_ancestors(
  File "/home/jk/tmp/venv/lib/python3.8/site-packages/tsinfer/inference.py", line 449, in match_ancestors
    return matcher.match_ancestors()
  File "/home/jk/tmp/venv/lib/python3.8/site-packages/tsinfer/inference.py", line 1401, in match_ancestors
    ts = self.store_output()
  File "/home/jk/tmp/venv/lib/python3.8/site-packages/tsinfer/inference.py", line 1477, in store_output
    ts = self.get_ancestors_tree_sequence()
  File "/home/jk/tmp/venv/lib/python3.8/site-packages/tsinfer/inference.py", line 1448, in get_ancestors_tree_sequence
    self.convert_inference_mutations(tables)
  File "/home/jk/tmp/venv/lib/python3.8/site-packages/tsinfer/inference.py", line 1232, in convert_inference_mutations
    metadata = _update_site_metadata(site.metadata, constants.INFERENCE_FULL)
  File "/home/jk/tmp/venv/lib/python3.8/site-packages/tsinfer/inference.py", line 150, in _update_site_metadata
    return {"inference_type": inference_type, **current_metadata}
TypeError: 'int' object is not a mapping
```

Somehow we're setting the metadata to 0 for sites when I run the version from a pip installation, but not when working with the git version. Very strange...

**Comment by frezaei98** [comment]
> I can confirm this happens in a fresh venv... investigating.

I have been working on tsinfer for three or four months and it was very strange that it was right until Monday, but on Tuesday I want to continue my work, I got an error....

**Comment by frezaei98** [comment]
> Very strange:
> 
> ```
> (venv) jk@tempest$ python3 tsinfer-example.py 
> Traceback (most recent call last):
>   File "tsinfer-example.py", line 12, in <module>
>     inferred_ts = tsinfer.infer(sample_data)
>   File "/home/jk/tmp/venv/lib/python3.8/site-packages/tsinfer/inference.py", line 291, in infer
>     ancestors_ts = match_ancestors(
>   File "/home/jk/tmp/venv/lib/python3.8/site-packages/tsinfer/inference.py", line 449, in match_ancestors
>     return matcher.match_ancestors()
>   File "/home/jk/tmp/venv/lib/python3.8/site-packages/tsinfer/inference.py", line 1401, in match_ancestors
>     ts = self.store_output()
>   File "/home/jk/tmp/venv/lib/python3.8/site-packages/tsinfer/inference.py", line 1477, in store_output
>     ts = self.get_ancestors_tree_sequence()
>   File "/home/jk/tmp/venv/lib/python3.8/site-packages/tsinfer/inference.py", line 1448, in get_ancestors_tree_sequence
>     self.convert_inference_mutations(tables)
>   File "/home/jk/tmp/venv/lib/python3.8/site-packages/tsinfer/inference.py", line 1232, in convert_inference_mutations
>     metadata = _update_site_metadata(site.metadata, constants.INFERENCE_FULL)
>   File "/home/jk/tmp/venv/lib/python3.8/site-packages/tsinfer/inference.py", line 150, in _update_site_metadata
>     return {"inference_type": inference_type, **current_metadata}
> TypeError: 'int' object is not a mapping
> ```
> 
> Somehow we're setting the metadata to 0 for sites when I run the version from a pip installation, but not when working with the git version. Very strange...

**Comment by jeromekelleher** [comment]
I'm going to reopen this @frezaei98 - there's definitely something wrong here. I'm not sure if it's directly a bug in tsinfer or some weird interaction with an upstream package, but it's something we need to get to the bottom of.

**Comment by jeromekelleher** [comment]
I've tracked this down to a problem with Zarr @frezaei98. Something changed in the behaviour of metadata with version 2.11.

MWE:
```python
import tsinfer

with tsinfer.SampleData(sequence_length=6) as sample_data:
    sample_data.add_site(0, [0, 1, 0, 0, 0], ["A", "T"])

for site in sample_data.sites():
    print(site)
```

with version 2.10 we get:
```
Site(id=0, position=0.0, ancestral_state='A', metadata={}, time=nan, alleles=('A', 'T'))
```
and version 2.11 we have 
```
Site(id=0, position=0.0, ancestral_state='A', metadata=0, time=nan, alleles=('A', 'T'))
```

Note the ``metadata=0`` rather than ``metadata={}``.

The short term fix is to install zarr version 2.10, e.g. ``python3 -m pip install zarr==2.10``.

I'm not sure if this is a bug in zarr or whether we were depending on some undocumented/dodgy behaviour. I guess we need to figure this out before we decide what the right approach is.

**Comment by hyanwong** [comment]
Thanks for tracking this down @jeromekelleher. Good bug hunting there.

**Comment by benjeffery** [comment]
Bisected to https://github.com/zarr-developers/zarr-python/commit/f461eb78fbb88187582cd9123d6ec7622d9abd26 still figuring out which bit of it!

**Comment by hyanwong** [comment]
Good luck @benjeffery !

**Comment by jeromekelleher** [comment]
Aha, fantastic, thanks @benjeffery! I bet it's the ``write_empty_chunks=False`` default that's hit us, and we just have to set that to True explicitly. I'll open a PR to test.

**Comment by benjeffery** [comment]
> Aha, fantastic, thanks @benjeffery! I bet it's the `write_empty_chunks=False` default that's hit us, and we just have to set that to True explicitly. I'll open a PR to test.

Yep, I expect so. Was just trying to find the spot in tsinfer.

**Comment by jeromekelleher** [comment]
Yeah, I tried to do a quick fix over in #631, which didn't work. Thanks for following up!

**Comment by benjeffery** [comment]
Reported upstream https://github.com/zarr-developers/zarr-python/issues/965

**Comment by frezaei98** [comment]
> .



> I've tracked this down to a problem with Zarr @frezaei98. Something changed in the behaviour of metadata with version 2.11.
> 
> MWE:
> 
> ```python
> import tsinfer
> 
> with tsinfer.SampleData(sequence_length=6) as sample_data:
>     sample_data.add_site(0, [0, 1, 0, 0, 0], ["A", "T"])
> 
> for site in sample_data.sites():
>     print(site)
> ```
> 
> with version 2.10 we get:
> 
> ```
> Site(id=0, position=0.0, ancestral_state='A', metadata={}, time=nan, alleles=('A', 'T'))
> ```
> 
> and version 2.11 we have
> 
> ```
> Site(id=0, position=0.0, ancestral_state='A', metadata=0, time=nan, alleles=('A', 'T'))
> ```
> 
> Note the `metadata=0` rather than `metadata={}`.
> 
> The short term fix is to install zarr version 2.10, e.g. `python3 -m pip install zarr==2.10`.
> 
> I'm not sure if this is a bug in zarr or whether we were depending on some undocumented/dodgy behaviour. I guess we need to figure this out before we decide what the right approach is.

Many thanks. I changed the version of Zarr package and it fixed and worked.

**Comment by benjeffery** [comment]
Here's an additional zarr issue that we hit in this release:
https://github.com/zarr-developers/zarr-python/issues/967
